# Bit Manipulation
- [cp-algorithms - Bit Manipulation](https://cp-algorithms.com/algebra/bit-manipulation.html)
- [Striver Bit Manipulation Playlist](https://cp-algorithms.com/algebra/bit-manipulation.html)

## Decimal to base conversion
- convert 5 to binary (base 2) = 101
- convert 10 to octal (base 8) = 12
- Time complexity : log<sub>base</sub>(n), it depends on the number of steps we divide
- Space complexity : log<sub>base</sub>(n), we add the remainder as the number of the steps

In [35]:
fun getDecimalToBase(num: Int, base: Int): String {
    val sb = StringBuilder()
    var _num = num
    while (_num > 0) {
        sb.append(_num % base)
        _num = _num / base
    }
    return sb.reverse().toString()
}

fun getBinaryToDecimal(num: String, base: Int): Int {
    var result = 0
    var power = 1
    for (i in num.length - 1 downTo 0) {
        val digit = num[i] - '0'
        result += digit * power
        power *= base
    }

    return result
}

In [36]:
println("5 to its binary form : ${getDecimalToBase(5, 2)}")
println("10 to its octal form : ${getDecimalToBase(10, 8)}")
println("10 to its binary form : ${getDecimalToBase(10, 2)}")

// Reverse
println("1010 (binary) to its decimal form : ${getBinaryToDecimal("1010", 2)}")
println("12 (octal) to its decimal form : ${getBinaryToDecimal("12", 8)}")
println("1111 (binary) to its decimal form : ${getBinaryToDecimal("1111", 2)}")

5 to its binary form : 101
10 to its octal form : 12
10 to its binary form : 1010
1010 (binary) to its decimal form : 10
12 (octal) to its decimal form : 10
1111 (binary) to its decimal form : 15


 ### Kotlin tricks to convert to any radix and hex format and reverse

In [22]:
@file:OptIn(ExperimentalStdlibApi::class)

println("5 to its binary form : ${5.toString(2)}")
println("10 to its octal form : ${10.toString(8)}")
println("10 to its binary form : ${10.toString(2)}")
println("10 to its binary form : ${10.toHexString()}")
println("1010 (binary) to its decimal form : ${"1010".toInt(2)}")
println("12 (octal) to its decimal form : ${"12".toInt(8)}")

5 to its binary form : 101
10 to its octal form : 12
10 to its binary form : 1010
10 to its binary form : 0000000a
1010 (binary) to its decimal form : 10
12 (octal) to its decimal form : 10


## Operators (AND, OR, XOR, SHIFT, NOT)
- Left shift is multiply by base. (2^31 - 1 if left shifted, then will **overflow**)
- Right shift is dividing by base

In [66]:
val num1 = 13
val num2 = 7

println("And Operation of ${num1.toString(2)} and ${num2.toString(2)} is ${num1 and num2}")
println("OR Operation of ${num1.toString(2)} and ${num2.toString(2)} is ${num1 or num2}")
println("XOR Operation of ${num1.toString(2)} and ${num2.toString(2)} is ${num1 xor num2}")
println("Right Shift of ${num1.toString(2)} is ${(num1 shr 1).toString(2)}")
println("Left Shift of ${num1.toString(2)} is ${(num1 shl 1).toString(2)}")
println("Integer max overflow is : ${Integer.MAX_VALUE shl 1}") // 11111111111111111111111111111110 (-2) in two complement

println("NOT of ${num1.toString(2)} is ${(num1.inv().toString(2))}")
// val num = 13            // 00000000 00000000 00000000 00001101
// val flipped = num.inv() // 11111111 11111111 11111111 11110010 (which is -14 in 2's complement)
// println(flipped)


And Operation of 1101 and 111 is 5
OR Operation of 1101 and 111 is 15
XOR Operation of 1101 and 111 is 10
Right Shift of 1101 is 110
Left Shift of 1101 is 11010
Integer max overflow is : -2
NOT of 1101 is -1110


## Max value of Integer and how the 32 bits in integer are stored
- Negative numbers : Convert number in 1s complement and then add 1. Flip the last bit
- For storing -1, doesn't use the last signed bit, it converts 1 to flip, then adds 1. (So, its 2^31, uses just the last bit)
- Max Value : +ve 2^31 - 1 (setting all its as 1 except last bit reserved for -ve)
- Min Value : -ve 2^31

In [116]:
fun Int.to32bitString(): String = Integer.toBinaryString(this).padStart(Int.SIZE_BITS, '0')

println(Int.MIN_VALUE.to32bitString() + " ${Integer.MIN_VALUE}")
println((-1).to32bitString() + " -1 ")
println(0.to32bitString() + " 0")
println(1.to32bitString() + " 1")
println(Int.MAX_VALUE.to32bitString() + " ${Integer.MAX_VALUE}")
println((-2).to32bitString() + " ${-2}")


10000000000000000000000000000000 -2147483648
11111111111111111111111111111111 -1 
00000000000000000000000000000000 0
00000000000000000000000000000001 1
01111111111111111111111111111111 2147483647
11111111111111111111111111111110 -2


### Swap two numbers
- Property of xor is to cancel if same.

In [74]:
var num1 = 5
var num2 = 6
println("Before swap : $num1, $num2")
num1 = num1 xor num2
num2 = num1 xor num2
num1 = num1 xor num2
println("After swap : $num1, $num2")

Before swap : 5, 6
After swap : 6, 5


### Check if ith bit is set

In [75]:
var num = 13
var i = 2
println("Is $i th bit set in ${num.toString(2)} : ${((num shr i) and 1) == 1}")
println("Is $i th bit set in ${num.toString(2)} : ${((1 shl i) and num) > 0}")
i = 1
println("Is $i th bit set in ${num.toString(2)} : ${((num shr i) and 1) == 1}")
println("Is $i th bit set in ${num.toString(2)} : ${((1 shl i) and num) > 0}")

Is 2 th bit set in 1101 : true
Is 2 th bit set in 1101 : true
Is 1 th bit set in 1101 : false
Is 1 th bit set in 1101 : false


### Set/Clear/Toggle the ith bit

In [85]:
// Set
var num = 5
var i = 1
println("Setting the 1st bit of $num : ${(1 shl i) or num}")
num = 13
i = 2
println("Setting the 1st bit of $num : ${(1 shl i) or num}")

// Clear
num = 5
i = 1
println("Clearing the 1st bit of $num : ${(1 shl i).inv() and num}")
num = 13
i = 2
println("Clearing the 1st bit of $num : ${(1 shl i).inv() and num}")


// Toggle
num = 5
i = 1
println("Before Toggling $num")
num = (1 shl i) xor num
println("Toggling once 1st bit $num")
num = (1 shl i) xor num
println("Toggling back 1st bit $num")

num = 13
i = 2
println("Before Toggling $num")
num = (1 shl i) xor num
println("Toggling once 1st bit $num")
num = (1 shl i) xor num
println("Toggling back 1st bit $num")

Setting the 1st bit of 5 : 7
Setting the 1st bit of 13 : 13
Clearing the 1st bit of 5 : 5
Clearing the 1st bit of 13 : 9
Before Toggling 5
Toggling once 1st bit 7
Toggling back 1st bit 5
Before Toggling 13
Toggling once 1st bit 9
Toggling back 1st bit 13


### Clearing the right-most set bit
- 0110 -> 0100
- 1100 -> 1000
- 1000 (8) and 7 is 0111. N and N-1 have a relation that we are going to use for solving this. (n and n-1) is always 0

In [89]:
var num = 6
println("after operation $num is ${num and num - 1}")
num = 12
println("after operation $num is ${num and num - 1}")

after operation 6 is 4
after operation 12 is 8


### Check if the number is power of 2

In [94]:
var num = 10
println("Is $num power of 2 : ${(num and num - 1) == 0}")
num = 16
println("Is $num power of 2 : ${(num and num - 1) == 0}")
num = 32
println("Is $num power of 2 : ${(num and num - 1) == 0}")
num = 31
println("Is $num power of 2 : ${(num and num - 1) == 0}")

Is 10 power of 2 : false
Is 16 power of 2 : true
Is 32 power of 2 : true
Is 31 power of 2 : false


### Count the number of set bits

#### Kotlin inbuilt method

In [96]:
// Kotlin in built method
println("Number of bits in 5 is ${5.countOneBits()}")

Number of bits in 5 is 2


#### Brute force O(LogN)

In [103]:
var num = 13
var count = 0
while (num > 0) {
    println("num : $num, count : $count")
    count += num and 1 // To know if the number is odd check the last bit will be set to 1
    num = num shr 1
}
println("Count : $count")

num : 13, count : 0
num : 6, count : 1
num : 3, count : 1
num : 1, count : 2
Count : 3


#### Optimized
- Time Complexity : O(31 N ): Right most bit will be set 0, will be running till all the bits are set 0. Max bits in integer can 31

In [105]:
var num = 13
var count = 0
while (num > 0) {
    println("count : $count, num : $num")
    num = num and num - 1 // Right most bit will be set 0, will be running till all the bits are set 0
    count += 1
}
println("count : $count")

 count : 0, num : 13
 count : 1, num : 12
 count : 2, num : 8
count : 3


In [118]:
// Another way is to shift left 1 and check
fun countBits2(num: Int): Int {
    var count = 0
    repeat(31) {
        if (num and (1 shl it) > 0) {
            count += 1
        }
    }
    return count
}

var num = 13
println("No of ones in $num : ${countBits2(num)}")
num = Integer.MAX_VALUE
println("No of ones in $num : ${countBits2(num)}")

No of ones in 13 : 3
No of ones in 2147483647 : 31


### Check if a number is power of two

In [120]:
var x = 16
var result = (x > 0) && (x and (x - 1) == 0)
println("Is $x power of two : $result")

x = 15
result = (x > 0) && (x and (x - 1) == 0)
println("Is $x power of two : $result")


Is 16 power of two : true
Is 15 power of two : false


### Isolate the rightmost set bit

In [125]:
var x = 10
println("Right most set in ${x.to32bitString()} bit is : ${(x and -x).to32bitString()}")
x = 1
println("Right most set in ${x.to32bitString()} bit is : ${(x and -x).to32bitString()}")
x = 24
println("Right most set in ${x.to32bitString()} bit is : ${(x and -x).to32bitString()}")

Right most set in 00000000000000000000000000001010 bit is : 00000000000000000000000000000010
Right most set in 00000000000000000000000000000001 bit is : 00000000000000000000000000000001
Right most set in 00000000000000000000000000011000 bit is : 00000000000000000000000000001000


## [191. Number of 1 Bits](https://leetcode.com/problems/number-of-1-bits/description/)
<pre>
Given a positive integer n, write a function that returns the number of set bits in its binary representation (also known as the Hamming weight).
Example 1:
Input: n = 11
Output: 3
Explanation: The input binary string 1011 has a total of three set bits.
</pre>

- Time Complexity :
- Space Complexity :


## [338. Counting Bits](https://leetcode.com/problems/counting-bits/description/)

<pre>
Given an integer n, return an array ans of length n + 1 such that for each i (0 <= i <= n), ans[i] is the number of 1's in the binary representation of i.
Example 1:
Input: n = 2
Output: [0,1,1]
Explanation:
0 --> 0
1 --> 1
2 --> 10
</pre>

- Time Complexity :
- Space Complexity :

## [762.Prime Number of Set Bits in Binary Representation](https://leetcode.com/problems/prime-number-of-set-bits-in-binary-representation/description/)

<pre>
Given two integers left and right, return the count of numbers in the inclusive range [left, right] having a prime number of set bits in their binary representation.
Recall that the number of set bits an integer has is the number of 1's present when written in binary.
For example, 21 written in binary is 10101, which has 3 set bits.

Example 1:
Input: left = 6, right = 10
Output: 4
Explanation:
6  -> 110 (2 set bits, 2 is prime)
7  -> 111 (3 set bits, 3 is prime)
8  -> 1000 (1 set bit, 1 is not prime)
9  -> 1001 (2 set bits, 2 is prime)
10 -> 1010 (2 set bits, 2 is prime)
4 numbers have a prime number of set bits.
</pre>

- Time Complexity :
- Space Complexity :

## [1017. Convert to Base -2](https://leetcode.com/problems/convert-to-base-2/description/)
<pre>
Given an integer n, return a binary string representing its representation in base -2.
Note that the returned string should not have leading zeros unless the string is "0".
Example 1:

Input: n = 2
Output: "110"
Explantion: (-2)2 + (-2)1 = 2
</pre>

- Time Complexity :
- Space Complexity :

## [1009. Complement of Base 10 Integer](https://leetcode.com/problems/complement-of-base-10-integer/) / [476. Number Complement](https://leetcode.com/problems/number-complement/description/)
<pre>
The complement of an integer is the integer you get when you flip all the 0's to 1's and all the 1's to 0's in its binary representation.
For example, The integer 5 is "101" in binary and its complement is "010" which is the integer 2.
Given an integer n, return its complement.
Input: n = 5
Output: 2
Explanation: 5 is "101" in binary, with complement "010" in binary, which is 2 in base-10.
</pre>

